src/services/httpservice.cpp

```diff
@@ -91,17 +95,96 @@ void HTTPService::run(std::streambuf *in, std::streambuf *out, std::streambuf *e
-                }
+            response.send500(concat("invalid request: ", e.what()));
+        } else {
+            response.send500("invalid request");
+        }
 	}
 	Log::off();
 }
 
+void HTTPService::catchExceptions(HTTPResponseStream& response, const MappingException &me){
+    auto exception_type = me.getExceptionType();
+	const bool global_debug = Configuration::get<bool>("global.debug", false);
+
+	if(global_debug || exception_type != MappingExceptionType::CONFIDENTIAL){
+        Json::Value exceptionJson;
+        readNestedException(exceptionJson, me);
+
+        if(!global_debug){
+            bool all_confidential = clearExceptionJsonFromConfidential(exceptionJson);
+            if(all_confidential){
+                response.send500("invalid request");
+                return;
+            }
+        }
+        response.sendHeader("Status", "500 Internal Server Error");
+        response.sendJSON(exceptionJson);
+    } else {
+        response.send500("invalid request");
+    }
+}
+
+void HTTPService::readNestedException(Json::Value &exceptionJson, const MappingException &me){
+    auto type = me.getExceptionType();
+
+    try {
+        exceptionJson["message"] = me.what();
+        switch(type){
+            case MappingExceptionType::CONFIDENTIAL:
+                exceptionJson["type"] = "CONFIDENTIAL";
+                break;
+            case MappingExceptionType::PERMANENT:
+                exceptionJson["type"] = "PERMANENT";
+                break;
+            case MappingExceptionType::TRANSIENT:
+                exceptionJson["type"] = "TRANSIENT";
+                break;
+            case MappingExceptionType::SAME_AS_NESTED:
+                exceptionJson["type"] = "SAME_AS_NESTED";
+                break;
+        }
+
+        std::rethrow_if_nested(me);
+
+    } catch(MappingException &nested) {
+        readNestedException(exceptionJson["nested_exception"], nested);
+    } catch(std::runtime_error &e){
+        Json::Value error_obj;
+        error_obj["message"] = "non_mapping_exception";
+        error_obj["type"] = "non_mapping_exception";
+        exceptionJson["nested_exception"] = error_obj;
+    }
+}
+
+bool HTTPService::clearExceptionJsonFromConfidential(Json::Value &exceptionJson){
+
+    if(exceptionJson["type"].asString() == "CONFIDENTIAL")
+        return true;
+
+    const bool hasNested = exceptionJson.isMember("nested_exception");
+
+    if(exceptionJson["type"].asString() == "SAME_AS_NESTED"){
+        if(hasNested)
+            return clearExceptionJsonFromConfidential(exceptionJson["nested_exception"]);
+        else
+            return true; //SAME_AS_NESTED but no nested exception exists, so remove it just in case.
+    } else {
+        //TRANSIENT or PERMANENT as type. If a nested confidential exception exists, remove it.
+        const bool nestedIsConfidential = clearExceptionJsonFromConfidential(exceptionJson["nested_exception"]);
```

Comment:
doesn't it crash here if there is no nested exception?

Reply:
It does not, but only because jsoncpp handles not existing members without crashing and returns a "null" `Json::Value` instead. So I will adjust it.